Python/optimizer_analysis.c

```diff
@@ -43,6 +44,119 @@ _PyOpcode_isterminal(int opcode)
 }
 
 
+typedef enum {
+    PYINT_TYPE = 0,
+    PYFLOAT_TYPE = 1,
+    PYUNICODE_TYPE = 2,
+    // GUARD_GLOBALS_VERSION_TYPE,
+    // GUARD_BUILTINS_VERSION_TYPE,
+    GUARD_DORV_VALUES_TYPE = 3,
+    GUARD_TYPE_VERSION_STORE_TYPE = 4,
+    GUARD_DORV_VALUES_INST_ATTR_FROM_DICT_TYPE = 5,
+    GUARD_KEYS_VERSION_TYPE = 6,
+    GUARD_TYPE_VERSION_TYPE = 7,
+    // CHECK_CALL_BOUND_METHOD_EXACT_ARGS_TYPE, // idk how to deal with this, requires stack check
+    // CHECK_PEP_523_TYPE, // Environment check
+    // CHECK_FUNCTION_EXACT_ARGS_TYPE, // idk how to deal with this, requires stack check
+    // CHECK_STACK_SPACE_TYPE // Environment check
+    INVALID_TYPE = -1
+} _Py_UOpsSymExprTypeEnum;
+
+#define MAX_TYPE 7
+typedef struct {
+    // bitmask of types
+    uint32_t types;
+    // auxillary data for the types
+    uint32_t aux[MAX_TYPE + 1];
```

Comment:
Is this due to the fact that a single thing could be of multiple positive types?

Would it be possible to not have this, and only have something identify as a single positive type?

Reply:
Nope, its bcuz multiple guards has to pass for some cases inorder to fully eliminate the entire guard region. I dont see whats the disadvantage of doing this either

Reply:
Every time we add a new type, all sym expressions become 32 bits bigger. I would like to save memory somehow. But ok lets go with this for now.